src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages-with-web-push-libraries.md

```diff
@@ -0,0 +1,295 @@
+            return triggerPushMsg(subscription, dataToSend);
+          });
+        }
+
+        return promiseChain;
+      })
+
+The function `triggerPushMsg()` can then use the web-push library to send a message to the
+provided subscription.
+
+    const triggerPushMsg = function(subscription, dataToSend) {
+      return webpush.sendNotification(subscription, dataToSend)
+      .catch((err) => {
+        if (err.statusCode === 410) {
+          return deleteSubscriptionFromDatabase(subscription._id);
+        } else {
+          console.log('Subscription is no longer valid: ', err);
+        }
+      });
+    };
+
+The call to `webpush.sendNotification()` will return a promise. If the message was sent
+successfully the promise will resolves and there is nothing we need to do. If the promise
+rejects, we need to handle the error, which may involve remove the subscription from our data
+(i.e. the subscription is invalid) or perhaps the push service just could not fulfill our
+request and we need to try again later.
+
+To determine the type of error from a push service it's best to look at the status code. Error
+messages vary between push services and some are more helpful than others.
+
+In this example it checks for status code '404' and '410', which are the HTTP status codes for
+'Not Found' and 'Gone'. If we receive this status code, it means the subscription has expired
+or is no longer valid, in these scenarios we need remove the subscriptions from our database.
+
+We'll cover some of the other status codes in the next section when we look at the web push
+protocol in more detail.
+
+Note: If you hit problems at this stage, it's worth looking at the error logs from Firefox before
+Chrome. The Mozilla push service has much more helpful error messages compared to Chrome / FCM.
+
+After looping through the subscriptions, we need to return a JSON response.
+
+      .then(() => {
+        res.setHeader('Content-Type', 'application/json');
+          res.send(JSON.stringify({ data: { success: true } }));
+      })
+      .catch(function(err) {
+        res.status(500);
+        res.setHeader('Content-Type', 'application/json');
+        res.send(JSON.stringify({
+          error: {
+            id: 'unable-to-send-messages',
+            message: `We were unable to send messages to all subscriptions : ` +
+              `'${err.message}'`
+          }
+        }));
+      });
+
+We've gone over the major implementation steps.
+
+1. Create an API to send subscriptions from our web page to our back-end so it can save them in
+a database.
+1. Created an API to trigger sending push messages (in this case an API called from the pretend
+admin panel).
+1. Looked at an example of how to get all the subscriptions from our backend
+and send a message to each one with one of the [web-push
+libraries](https://github.com/web-push-libs/).
+
+Regardless of your backend (Node, PHP, Python, ....) the steps for implementing push are going
```

Comment:
Replace '....' (four periods) with either '...' (three periods, the proper form) or 'etc.'.